{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/importlib/_bootstrap.py:321: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  return f(*args, **kwds)\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(\"\"))))\n",
    "from sequence_annotation.utils.exception import InvalidStrandType,NegativeNumberException\n",
    "from sequence_annotation.utils.exception import ReturnNoneException\n",
    "from sequence_annotation.genome_handler.seq_info_parser import USCUParser\n",
    "from sequence_annotation.genome_handler.ann_genome_creator import AnnGenomeCreator\n",
    "from sequence_annotation.genome_handler.region_extractor import RegionExtractor\n",
    "from sequence_annotation.genome_handler.seq_info_generator import SeqInfoGenerator\n",
    "from sequence_annotation.genome_handler.seq_container import SeqInfoContainer\n",
    "genome_information={'chromosome':{'chr1':30,'chr2':35},'source':'unit_test'}\n",
    "file_prefix = \"../test/data/ucsc/\"\n",
    "principle = {'remove_end_of_strand':True,'with_random_choose':True,'replaceable':True,\n",
    "             \"each_region_number\":{'intergenic_region':10,'cds':100},\n",
    "             'sample_number_per_region':6,'half_length':4,'max_diff':3\n",
    "            }\n",
    "chroms_info = {'chr1':120,'chr2':35}\n",
    "frontground_types = ['cds','intron','utr_5','utr_3']\n",
    "background_type = 'intergenic_region'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Following regions won't be used to generate data:\n",
      "{'extra_index': None, 'id': 'region_chr1_minus_0', 'end': 29, 'ann_type': 'intergenic_region', 'note': '', 'source': 'unit_test', 'start': 0, 'ann_status': 'whole', 'strand': 'minus', 'extra_index_name': None, 'chromosome_id': 'chr1'}\n",
      "{'extra_index': None, 'id': 'region_chr1_plus_0', 'end': 3, 'ann_type': 'intergenic_region', 'note': '', 'source': 'unit_test', 'start': 0, 'ann_status': 'whole', 'strand': 'plus', 'extra_index_name': None, 'chromosome_id': 'chr1'}\n",
      "{'extra_index': None, 'id': 'region_chr2_minus_0', 'end': 34, 'ann_type': 'intergenic_region', 'note': '', 'source': 'unit_test', 'start': 0, 'ann_status': 'whole', 'strand': 'minus', 'extra_index_name': None, 'chromosome_id': 'chr2'}\n",
      "{'extra_index': None, 'id': 'region_chr2_plus_0', 'end': 34, 'ann_type': 'intergenic_region', 'note': '', 'source': 'unit_test', 'start': 0, 'ann_status': 'whole', 'strand': 'plus', 'extra_index_name': None, 'chromosome_id': 'chr2'}\n"
     ]
    }
   ],
   "source": [
    "file_path = file_prefix + 'one_plus_strand_both_utr.tsv'\n",
    "parser = USCUParser(file_path)\n",
    "parser.parse()\n",
    "creator = AnnGenomeCreator(genome_information,parser.data)\n",
    "creator.create()\n",
    "chroms = SeqInfoContainer()\n",
    "for chrom in creator.data.data:\n",
    "    extractor = RegionExtractor(chrom,frontground_types,background_type)\n",
    "    extractor.extract()\n",
    "    chroms.add(extractor.regions.data)\n",
    "generator = SeqInfoGenerator(chroms,principle,chroms_info,\"seed\",\"seq\")\n",
    "generator.generate()\n",
    "generator.seeds.to_data_frame().to_csv(\"seed.csv\")\n",
    "generator.seqs_info.to_data_frame().to_csv(\"seq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = generator.seqs_info.to_data_frame()\n",
    "selected_df = df[['id','source','strand']].copy()\n",
    "selected_df['start'] = df['start'] + 1\n",
    "selected_df['end'] = df['end'] + 1\n",
    "selected_df['feature'] = df['ann_type']\n",
    "selected_df['score'] = '.'\n",
    "selected_df['frame'] = '.'\n",
    "selected_df['attribute'] = '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_df['strand']=selected_df['strand'].str.replace(\"plus\", '+')\n",
    "selected_df['strand']=selected_df['strand'].str.replace(\"minus\", '-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>source</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature</th>\n",
       "      <th>score</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq_0</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_0</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq_1</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_0</td>\n",
       "      <td>+</td>\n",
       "      <td>17</td>\n",
       "      <td>26</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq_10</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_1</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>27</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq_100</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq_101</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>seq_102</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>seq_103</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>seq_104</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>seq_105</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>seq_106</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>seq_107</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_15</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>seq_108</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>seq_109</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>seq_11</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_1</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>25</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>seq_110</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>25</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>seq_111</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>seq_112</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>seq_113</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_16</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>25</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>seq_114</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>seq_115</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>seq_116</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>seq_117</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>seq_118</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>seq_119</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_17</td>\n",
       "      <td>+</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>seq_12</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_10</td>\n",
       "      <td>+</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>seq_120</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_18</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>seq_121</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_18</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>seq_122</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_18</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>seq_123</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_18</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>seq_124</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_18</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>seq_72</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>seq_73</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>seq_74</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>seq_75</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>seq_76</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>seq_77</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_109</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636</th>\n",
       "      <td>seq_78</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>637</th>\n",
       "      <td>seq_79</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>seq_8</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_1</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>25</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>seq_80</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>640</th>\n",
       "      <td>seq_81</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>seq_82</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>seq_83</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_11</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>seq_84</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>seq_85</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>seq_86</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>seq_87</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>seq_88</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>17</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>seq_89</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_12</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>seq_9</td>\n",
       "      <td>unit_test_region_chr1_plus_1_seed_1</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>seq_90</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>18</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>seq_91</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>seq_92</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>seq_93</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>seq_94</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>seq_95</td>\n",
       "      <td>unit_test_region_chr1_plus_5_seed_13</td>\n",
       "      <td>+</td>\n",
       "      <td>7</td>\n",
       "      <td>16</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>seq_96</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>seq_97</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>seq_98</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>11</td>\n",
       "      <td>24</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>seq_99</td>\n",
       "      <td>unit_test_region_chr1_plus_6_seed_14</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>cds</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>660 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id                                 source strand  start  end  \\\n",
       "0      seq_0    unit_test_region_chr1_plus_1_seed_0      +     16   28   \n",
       "1      seq_1    unit_test_region_chr1_plus_1_seed_0      +     17   26   \n",
       "2     seq_10    unit_test_region_chr1_plus_1_seed_1      +     16   27   \n",
       "3    seq_100   unit_test_region_chr1_plus_6_seed_14      +     14   23   \n",
       "4    seq_101   unit_test_region_chr1_plus_6_seed_14      +     11   22   \n",
       "5    seq_102   unit_test_region_chr1_plus_5_seed_15      +      1   15   \n",
       "6    seq_103   unit_test_region_chr1_plus_5_seed_15      +      3   13   \n",
       "7    seq_104   unit_test_region_chr1_plus_5_seed_15      +      2   13   \n",
       "8    seq_105   unit_test_region_chr1_plus_5_seed_15      +      4   15   \n",
       "9    seq_106   unit_test_region_chr1_plus_5_seed_15      +      3   14   \n",
       "10   seq_107   unit_test_region_chr1_plus_5_seed_15      +      4   13   \n",
       "11   seq_108   unit_test_region_chr1_plus_6_seed_16      +     12   22   \n",
       "12   seq_109   unit_test_region_chr1_plus_6_seed_16      +     11   23   \n",
       "13    seq_11    unit_test_region_chr1_plus_1_seed_1      +     16   25   \n",
       "14   seq_110   unit_test_region_chr1_plus_6_seed_16      +     14   25   \n",
       "15   seq_111   unit_test_region_chr1_plus_6_seed_16      +     13   23   \n",
       "16   seq_112   unit_test_region_chr1_plus_6_seed_16      +     11   23   \n",
       "17   seq_113   unit_test_region_chr1_plus_6_seed_16      +     12   25   \n",
       "18   seq_114   unit_test_region_chr1_plus_6_seed_17      +      9   20   \n",
       "19   seq_115   unit_test_region_chr1_plus_6_seed_17      +     11   23   \n",
       "20   seq_116   unit_test_region_chr1_plus_6_seed_17      +      9   23   \n",
       "21   seq_117   unit_test_region_chr1_plus_6_seed_17      +     11   21   \n",
       "22   seq_118   unit_test_region_chr1_plus_6_seed_17      +      9   22   \n",
       "23   seq_119   unit_test_region_chr1_plus_6_seed_17      +     10   22   \n",
       "24    seq_12   unit_test_region_chr1_plus_6_seed_10      +     10   24   \n",
       "25   seq_120   unit_test_region_chr1_plus_5_seed_18      +      5   13   \n",
       "26   seq_121   unit_test_region_chr1_plus_5_seed_18      +      2   13   \n",
       "27   seq_122   unit_test_region_chr1_plus_5_seed_18      +      3   15   \n",
       "28   seq_123   unit_test_region_chr1_plus_5_seed_18      +      2   13   \n",
       "29   seq_124   unit_test_region_chr1_plus_5_seed_18      +      3   15   \n",
       "..       ...                                    ...    ...    ...  ...   \n",
       "630   seq_72  unit_test_region_chr1_plus_6_seed_109      +     12   21   \n",
       "631   seq_73  unit_test_region_chr1_plus_6_seed_109      +     12   21   \n",
       "632   seq_74  unit_test_region_chr1_plus_6_seed_109      +     11   24   \n",
       "633   seq_75  unit_test_region_chr1_plus_6_seed_109      +     10   24   \n",
       "634   seq_76  unit_test_region_chr1_plus_6_seed_109      +     10   22   \n",
       "635   seq_77  unit_test_region_chr1_plus_6_seed_109      +     11   24   \n",
       "636   seq_78   unit_test_region_chr1_plus_5_seed_11      +      2   14   \n",
       "637   seq_79   unit_test_region_chr1_plus_5_seed_11      +      3   14   \n",
       "638    seq_8    unit_test_region_chr1_plus_1_seed_1      +     16   25   \n",
       "639   seq_80   unit_test_region_chr1_plus_5_seed_11      +      2   14   \n",
       "640   seq_81   unit_test_region_chr1_plus_5_seed_11      +      2   14   \n",
       "641   seq_82   unit_test_region_chr1_plus_5_seed_11      +      3   16   \n",
       "642   seq_83   unit_test_region_chr1_plus_5_seed_11      +      3   16   \n",
       "643   seq_84   unit_test_region_chr1_plus_5_seed_12      +      6   18   \n",
       "644   seq_85   unit_test_region_chr1_plus_5_seed_12      +      4   16   \n",
       "645   seq_86   unit_test_region_chr1_plus_5_seed_12      +      7   18   \n",
       "646   seq_87   unit_test_region_chr1_plus_5_seed_12      +      5   16   \n",
       "647   seq_88   unit_test_region_chr1_plus_5_seed_12      +      7   17   \n",
       "648   seq_89   unit_test_region_chr1_plus_5_seed_12      +      6   16   \n",
       "649    seq_9    unit_test_region_chr1_plus_1_seed_1      +     14   26   \n",
       "650   seq_90   unit_test_region_chr1_plus_5_seed_13      +      5   18   \n",
       "651   seq_91   unit_test_region_chr1_plus_5_seed_13      +      5   15   \n",
       "652   seq_92   unit_test_region_chr1_plus_5_seed_13      +      4   16   \n",
       "653   seq_93   unit_test_region_chr1_plus_5_seed_13      +      4   17   \n",
       "654   seq_94   unit_test_region_chr1_plus_5_seed_13      +      7   16   \n",
       "655   seq_95   unit_test_region_chr1_plus_5_seed_13      +      7   16   \n",
       "656   seq_96   unit_test_region_chr1_plus_6_seed_14      +     12   24   \n",
       "657   seq_97   unit_test_region_chr1_plus_6_seed_14      +     11   25   \n",
       "658   seq_98   unit_test_region_chr1_plus_6_seed_14      +     11   24   \n",
       "659   seq_99   unit_test_region_chr1_plus_6_seed_14      +     12   23   \n",
       "\n",
       "               feature score frame attribute  \n",
       "0    intergenic_region     .     .         .  \n",
       "1    intergenic_region     .     .         .  \n",
       "2    intergenic_region     .     .         .  \n",
       "3                  cds     .     .         .  \n",
       "4                  cds     .     .         .  \n",
       "5                  cds     .     .         .  \n",
       "6                  cds     .     .         .  \n",
       "7                  cds     .     .         .  \n",
       "8                  cds     .     .         .  \n",
       "9                  cds     .     .         .  \n",
       "10                 cds     .     .         .  \n",
       "11                 cds     .     .         .  \n",
       "12                 cds     .     .         .  \n",
       "13   intergenic_region     .     .         .  \n",
       "14                 cds     .     .         .  \n",
       "15                 cds     .     .         .  \n",
       "16                 cds     .     .         .  \n",
       "17                 cds     .     .         .  \n",
       "18                 cds     .     .         .  \n",
       "19                 cds     .     .         .  \n",
       "20                 cds     .     .         .  \n",
       "21                 cds     .     .         .  \n",
       "22                 cds     .     .         .  \n",
       "23                 cds     .     .         .  \n",
       "24                 cds     .     .         .  \n",
       "25                 cds     .     .         .  \n",
       "26                 cds     .     .         .  \n",
       "27                 cds     .     .         .  \n",
       "28                 cds     .     .         .  \n",
       "29                 cds     .     .         .  \n",
       "..                 ...   ...   ...       ...  \n",
       "630                cds     .     .         .  \n",
       "631                cds     .     .         .  \n",
       "632                cds     .     .         .  \n",
       "633                cds     .     .         .  \n",
       "634                cds     .     .         .  \n",
       "635                cds     .     .         .  \n",
       "636                cds     .     .         .  \n",
       "637                cds     .     .         .  \n",
       "638  intergenic_region     .     .         .  \n",
       "639                cds     .     .         .  \n",
       "640                cds     .     .         .  \n",
       "641                cds     .     .         .  \n",
       "642                cds     .     .         .  \n",
       "643                cds     .     .         .  \n",
       "644                cds     .     .         .  \n",
       "645                cds     .     .         .  \n",
       "646                cds     .     .         .  \n",
       "647                cds     .     .         .  \n",
       "648                cds     .     .         .  \n",
       "649  intergenic_region     .     .         .  \n",
       "650                cds     .     .         .  \n",
       "651                cds     .     .         .  \n",
       "652                cds     .     .         .  \n",
       "653                cds     .     .         .  \n",
       "654                cds     .     .         .  \n",
       "655                cds     .     .         .  \n",
       "656                cds     .     .         .  \n",
       "657                cds     .     .         .  \n",
       "658                cds     .     .         .  \n",
       "659                cds     .     .         .  \n",
       "\n",
       "[660 rows x 9 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"asd\"[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
